### PR TITLE
Tolerate leading "~/" or "/"

### PR DESCRIPTION
**Pull request by jbagga**
Addresses #441

### CODE REVIEW

File: src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternBuilder.cs [hunk 1]
```diff
@@ -71,6 +71,11 @@ public RoutePattern Build()
 
         public static RoutePatternBuilder Create(string text)
         {
+            if (text.StartsWith("~/"))
+            {
+                return new RoutePatternBuilder() { RawText = text.Substring(1), };
+            }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternBuilder.cs]: The `RawText` should still reflect what the user typed in. I don't think we should have this special case.

File: src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternParser.cs [hunk 2]
```diff
@@ -26,20 +26,16 @@ internal static class RoutePatternParser
             Asterisk
         };
 
-        public static RoutePattern Parse(string pattern)
+        public static RoutePattern Parse(string originalPattern)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternParser.cs]: not digging the rename here

File: src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternParser.cs [hunk 3]
```diff
@@ -469,10 +466,21 @@ private static bool IsValidLiteral(TemplateParserContext context, string literal
             return true;
         }
 
-        private static bool IsInvalidPattern(string routeTemplate)
+        private static string PatternStartsWithSlashOrTildeSlash(string routePattern)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternParser.cs]: This should be have a smarter name like `TrimPrefix` or `HandleLeadingSlash`. Based on the naming of this I expect this to return a boolean

File: src/Microsoft.AspNetCore.Dispatcher/Resources.resx [hunk 4]
```diff
@@ -164,7 +164,7 @@
     <value>The route parameter name '{0}' is invalid. Route parameter names must be non-empty and cannot contain these characters: '{{', '}}', '/'. The '?' character marks a parameter as optional, and can occur only at the end of the parameter. The '*' character marks a parameter as catch-all, and can occur only at the start of the parameter.</value>
   </data>
   <data name="TemplateRoute_InvalidRouteTemplate" xml:space="preserve">
-    <value>The route template cannot start with a '/' or '~' character.</value>
+    <value>The route template cannot start with a '~' character.</value>
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Dispatcher/Resources.resx]: This is a bit misleading, the route template can start with `~` if its followed by a `/`

Comment (jbagga) [reply src/Microsoft.AspNetCore.Dispatcher/Resources.resx]: `The route template cannot start with a '~' character unless followed by a "/".`

Is this any better?

File: src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternParser.cs (hunk 2 above)
Comment (jbagga) [reply src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternParser.cs]: Is it okay if I use the `trimmedPattern` after trimming the prefix and call this `pattern` again? I just need to make the distinction between the pattern that's the parameter and the one returned after trimming the leading '/' or "~/".

Comment (rynowak) [reply src/Microsoft.AspNetCore.Dispatcher/Patterns/RoutePatternParser.cs]: Sure, that sounds good.

**Comment by jbagga** [comment]
@rynowak Updated

**Pull request merged**